engine/orchestrator/score-orchestrator-impl/src/main/java/io/cloudslang/orchestrator/services/OrchestratorDispatcherServiceImpl.java

```diff
@@ -47,70 +41,107 @@
-        }
-        //This is done in order to make sure that we are not getting messages from worker that was already recovered and does not know about it yet
-        else if(!currentWRV.equals(wrv)){
-            logger.warn("Orchestrator got messages from worker: " + workerUuid + " with wrong WRV:" + wrv + " Current WRV is: " + currentWRV +  ". Discarding...");
-        }
-        else {
-            dispatch(messages);
-            workerNodeService.updateBulkNumber(workerUuid, bulkNumber);
+    public void dispatch(final List<? extends Serializable> messages, final String bulkNumber, String wrv, final String workerUuid) {
+        try {
+            //lock to synchronize with the recovery job
+            workerLockService.lock(workerUuid);
+            Validate.notNull(messages, "Messages list is null");
+
+            String currentBulkNumber = workerNodeService.readByUUID(workerUuid).getBulkNumber();
+            //can not be null at this point
+            String currentWRV = workerNodeService.readByUUID(workerUuid).getWorkerRecoveryVersion();
+
+            //This is done in order to make sure that if we do retries in worker we won't insert same bulk twice
+            if (currentBulkNumber != null && currentBulkNumber.equals(bulkNumber)) {
+                logger.warn("Orchestrator got messages bulk with same bulk number: " + bulkNumber + " This bulk was inserted to DB before. Discarding...");
+            }
+            //This is done in order to make sure that we are not getting messages from worker that was already recovered and does not know about it yet
+            else if (!currentWRV.equals(wrv)) {
+                logger.warn("Orchestrator got messages from worker: " + workerUuid + " with wrong WRV:" + wrv + " Current WRV is: " + currentWRV + ". Discarding...");
+            }
+            else {
+                dispatchBulk(messages, bulkNumber, workerUuid);
+            }
+        } catch (Exception ex) {
+            logger.error("Failed to dispatch bulk of messages: ", ex);
+            throw ex;
         }
     }
 
-    private void dispatch(List<? extends Serializable> messages) {
-        Validate.notNull(messages, "Messages list is null");
-
-        if (logger.isDebugEnabled()) logger.debug("Dispatching " + messages.size() + " messages");
-        long t = System.currentTimeMillis();
-        final AtomicInteger messagesCounter = new AtomicInteger(0);
-
-        dispatch(messages, ExecutionMessage.class, new Handler<ExecutionMessage>() {
-            @Override
-            public void handle(List<ExecutionMessage> messages) {
-                messagesCounter.addAndGet(messages.size());
-                queueDispatcher.dispatch(messages);
+    private void dispatchBulk(List<? extends Serializable> messages, String bulkNumber, String workerUuid){
+        try {
+            for(int i=0; i<3; i++){
```

Comment:
Done

From: Eliya Sadan [mailto:notifications@github.com]
Sent: Thursday, June 11, 2015 8:33 AM
To: CloudSlang/score
Cc: Beck, Natasha
Subject: Re: [score] Fault tolerance feature and UC on the queue table (#45)


In engine/orchestrator/score-orchestrator-impl/src/main/java/io/cloudslang/orchestrator/services/OrchestratorDispatcherServiceImpl.java<https://github.com/CloudSlang/score/pull/45#discussion_r32191034>:

>          }

>      }

>

> -    private void dispatch(List<? extends Serializable> messages) {

> -        Validate.notNull(messages, "Messages list is null");

> -

> -        if (logger.isDebugEnabled()) logger.debug("Dispatching " + messages.size() + " messages");

> -        long t = System.currentTimeMillis();

> -        final AtomicInteger messagesCounter = new AtomicInteger(0);

> -

> -        dispatch(messages, ExecutionMessage.class, new Handler<ExecutionMessage>() {

> -            @Override

> -            public void handle(List<ExecutionMessage> messages) {

> -                messagesCounter.addAndGet(messages.size());

> -                queueDispatcher.dispatch(messages);

> +    private void dispatchBulk(List<? extends Serializable> messages, String bulkNumber, String workerUuid){

> +        try {

> +            for(int i=0; i<3; i++){

please expose constants

—
Reply to this email directly or view it on GitHub<https://github.com/CloudSlang/score/pull/45/files#r32191034>.